### ISSUE TITLE
How to accept a completion item in an integration test? [dev question]

### ISSUE THREAD

**Issue by DanTup** [Microsoft/vscode#70320]
There's a lot of logic that goes on in the VS Code-side of things for completions, but I can't find a way to write an integration test like:

- Invoke completion
- Select an item
- Verify the document text changed in the expected way

Specifically, I can't see a way to apply/accept a completion item. I would like to ensure the completion items I am building are interpreted in the way I expect. There's logic that goes in inside VS code that isn't as simple as just applying the edit/additionalEdits - for example, the indentation [can go bad](https://github.com/Microsoft/vscode/issues/44200).

Is there a way to do this already? If not, can this be considered?

**Comment by jrieken** [comment]
> Is there a way to do this already? If not, can this be considered?

There is no way to do that, you could try executing the commands the UI executes but you will be on your own.

**Comment by DanTup** [comment]
Understood! I was able to make it work in a limited form by calling `editor.action.triggerSuggest` then `acceptSelectedSuggestion`. I can't find an obvious way to select a specific item (besides making it happen to be the top item by typing len-1 characters of the item I want) but that's not a deal breaker - I at least have some coverage that things are working together correctly :-) 

```ts
it.only("can complete unimported symbols and insert imports automatically", async () => {
		await setTestContent(`
main() {
  ProcessInf
}
		`);
		select(rangeOf("ProcessInf||"));

		await vs.commands.executeCommand("editor.action.triggerSuggest");
		await waitForSuggestionToResolve();
		await waitForEditorChange(() => vs.commands.executeCommand("acceptSelectedSuggestion"));
		await ensureTestContentWithCursorPos(`
import 'dart:io';

main() {
  ProcessInfo^
}
		`);
	});
```

I did discover https://github.com/Microsoft/vscode/issues/70374 while testing it though. Seems like tab completion will always insert the completion item immediately - even if it's not resolved. I don't know if that's expected - I can see arguments both ways (it can insert the wrong thing, but waiting for resolve would make pressing the tab key take time), but I'll leave that for you.
